gobblin-core/src/main/java/gobblin/source/extractor/extract/QueryBasedExtractor.java

```diff
@@ -166,6 +172,72 @@ public D readRecord(@Deprecated D reuse) throws DataRecordException, IOException
   }
 
   /**
+   * Check if it's appropriate to remove data pull upper bounds in the last work unit, fetching as much data as possible
+   * from the source. As between the time when data query was created and that was executed, there might be some
+   * new data generated in the source. Removing the upper bounds will help us grab the new data. It might result in a
+   * conservative high watermark, which is mitigated by {@link #setEstimatedActualHighWatermark()}
+   *
+   * Note: It's expected that there might be some duplicate data between runs because of removing the upper bounds
+   *
+   * @return should remove or not
+   */
+  private boolean shouldRemoveDataPullUpperBounds() {
+    // Only consider the last work unit
+    if (!this.workUnit.getPropAsBoolean(QueryBasedSource.IS_LAST_WORK_UNIT)) {
+      return false;
+    }
+
+    // Don't remove if the run is for 'APPEND'
+    if (!this.isFullDump()) {
+      return false;
+    }
+
+    // Don't remove if user specifies one or is recorded in previous run
+    if (this.workUnitState.getProp(ConfigurationKeys.SOURCE_QUERYBASED_END_VALUE) != null ||
+        this.workUnitState.getProp(ConfigurationKeys.WORK_UNIT_STATE_ACTUAL_HIGH_WATER_MARK_KEY) != null) {
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Remove all upper bounds in the predicateList used for pulling data
+   */
+  private void removeDataPullUpperBounds() {
+    log.info("Removing data pull upper bound for last work unit");
+    Iterator<Predicate> it = predicateList.iterator();
+    while (it.hasNext()) {
+      Predicate predicate = it.next();
+      if (predicate.getType() == Predicate.PredicateType.HWM) {
+        log.info("Remove predicate: " + predicate.condition);
+        it.remove();
+      }
+    }
+  }
+
+  /**
+   * Set an estimated actual high watermark when the work unit is the last one
+   * and its upper bounds to fetch data are removed
+   */
+  private void setEstimatedActualHighWatermark() {
```

Comment:
I think this function should be removed. Our local timestamp may not be in sync with the remote source, so this is not safe. The this.highWatermark value when computed from the source will give a timestamp that is appropriate for the source. Since we are removing the upper bound on the last partition, we may pull more data relative to this computed high watermark. We can handle dedup of duplicates when merging snapshots and appends.